### PR TITLE
Bump RN CLI, add @react-native/metro-config to template

### PR DESCRIPTION
**Pull request by huntie**
Summary:
Changelog:
[General][Changed] - The default `metro.config.js` in apps now extends `@react-native/metro-config`, and should be updated in existing apps.

### React Native Metro config → React Native repo (https://github.com/facebook/react-native/pull/36502)

We (the React Native team) are aiming to relocate the default Metro config for React Native out of `react-native-community/cli-plugin-metro` and **into the React Native repo + app template** as a new `react-native/metro-config` package.

This is the first (and minimum viable) phase we can ship to separate the release process of Metro from RN CLI in order to reduce coupling and iterate faster for our users.

**See full motivation, design, and test plan (which previewed the CLI bump) here: https://github.com/facebook/react-native/pull/36502**

## Changes

**This is part 2/2 of the @react-native/metro-config changes**, now that https://github.com/react-native-community/cli/pull/1875 is merged.

- Upgrade `react-native-community/cli` to `11.0.0`, upgrade all `metro` packages to `0.76.0` (version distributed in this CLI release).
- Remove hotfix https://github.com/facebook/react-native/pull/36631.
- Update the `metro.config.js` file in `packages/react-native/template/`.
    - Now merges defaults from `@react-native/metro-config`, and can be used with CLI >= 11.0.0.
- Update the `metro.config.js` files for `packages/react-native/` and `packages/rn-tester/` (these are integration test locations).
    - Now merges defaults from `@react-native/metro-config`, and can be used with CLI >= 11.0.0.

Changes to `@react-native/metro-config` — `0.72.1` (prepared but not depended on yet):

- Export `mergeConfig` util (removing direct `metro-config` dependency in consuming projects).
- Explicitly depend on `metro-react-native-babel-transformer` and `metro-runtime` (transitively included today).

Reviewed By: cortinico, blakef

Differential Revision: D44099691

### CODE REVIEW

**Comment by huntie** [comment]
🛠️ **This commit should resolve failing CircleCI tests on `main`.**

This appears to have happened because we expected to have bumped RN CLI to `11.0.0` in this commit — however we'd already been using the version specifier `"11.0.0-alpha.2"`, _where Yarn globs the portion of the specifier after `'-'`_.

https://github.com/facebook/react-native/blob/8c9c8ba5adb59f7f891a5307a0bce7200dd3ac7d/packages/react-native/package.json#L82

(Incidentally, [v11.0.0-alpha.1](https://github.com/react-native-community/cli/releases/tag/v11.0.0-alpha.1) was the most recent prerelease — `v11.0.0-alpha.2` had been unpublished from npm.)

Since the `template/` folder does not have a `yarn.lock`, and due to the interpretation of this specifier, when the RN CLI release was published yesterday .

---

Technically, a subset of this commit could have been merged earlier this week, that would have implemented the ["this stack, before new CLI is merged" test case](https://github.com/facebook/react-native/pull/36502) — and therefore have been forward-compatible in the event of an automatic bump, however I was waiting on:
- https://github.com/facebook/react-native/pull/36582

---

TLDR; the automatic major bump of RN CLI on `main` was not expected, and we should see what we can do to bring lockfile-equivalent behaviour to CI. cc @cortinico @cipolleschi

**Comment by NickGerleman** [comment]
> (Note: I understand why we don't have a lockfile in the template, but this is inherently variable for CI!)

I agree we should avoid anything non-deterministic where we can. There is also value in testing the real flow of users building a new app though. E.g. if this was a change in the NPM ecosystem which broke more RN versions than 0.72, folks trying to generate a new app, without an existing lockfile, could be broken the same way.

I think it is better to do something deterministic for diffs/PRs, but I think there would be value in continuing to test live dependencies in some continuous test.

**Comment by NickGerleman** [comment]
For where we do want determinism, could we just copy the repo lockfile into the template and do a yarn install? Any dependencies unique to the template would still be live, but then all of the ones defined in the main RN repo would be used otherwise.

**Review by cipolleschi — APPROVED** [review]

**Comment by huntie** [comment]
Because I'd temporarily deleted my `react-native` fork, GitHub isn't updating the "Files changed" tab or triggering CircleCI on this PR. https://github.com/facebook/react-native/pull/36697 is a copy of this PR to verify CircleCI stability. The Meta-internal commit is the source of truth.

**All Tests are passing for https://github.com/huntie/react-native/commit/ad7c99f957df9889bc4063031d40904dc502662e**:

image

https://github.com/facebook/react-native/pull/36697

- We still have some internal CI quirks to resolve in order to merge.
- In the meantime, @cipolleschi cherry picked all contents of this PR onto `0.72-rc` yesterday.

**Comment by huntie** [comment]
🎉
